File: src/tools/database_cluster.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { GraphQLClient } from 'graphql-request';
import { z } from 'zod';
import {
  GetCloudDatabaseClustersDocument,
  GetCloudDatabaseClusterDocument,
  CloudDatabaseClusterCreateDocument,
  CloudDatabaseClusterModifyDocument,
  CloudDatabaseClusterDeleteDocument,
  ClusterPlansDocument,
  ClusterVersionsDocument,
  GetCloudDatabaseClusterUserCredentialsDocument,
  type GetCloudDatabaseClustersQuery,
  type GetCloudDatabaseClusterQuery,
  type CloudDatabaseClusterCreateMutation,
  type CloudDatabaseClusterModifyMutation,
  type ClusterPlansQuery,
  type ClusterVersionsQuery,
  type GetCloudDatabaseClusterUserCredentialsQuery,
} from '../generated/graphql.js';

const AllowListEntry = z.object({
  ip: z.string(),
  state: z.enum(['PRESENT', 'ABSENT']),
});

const ExternalConnectionPort = z.object({
  externalPort: z
    .number()
    .optional()
    .describe(
      'Omit when creating a new port — the platform assigns the external port automatically. Required when modifying or removing an existing port so the backend knows which port to target.',
    ),
  internalPort: z.number().optional(),
  protocol: z.enum(['TCP', 'UDP']),
  state: z.enum(['PRESENT', 'ABSENT']),
  allowList: z.array(AllowListEntry),
});

const ExternalConnection = z.object({
  sharedIp: z.boolean(),
  state: z.enum(['PRESENT', 'ABSENT']),
  ports: z.array(ExternalConnectionPort),
});

const DatabaseInput = z.object({
  name: z.string(),
  state: z.enum(['PRESENT', 'ABSENT']),
  description: z.string().optional(),
});

const DatabaseUserPermission = z.object({
  databaseName: z.string(),
  permission: z.enum(['READ_ONLY', 'READ_WRITE']),
  state: z.enum(['PRESENT', 'ABSENT']),
});

const DatabaseUserInput = z.object({
  name: z.string(),
  password: z.string().optional(),
  state: z.enum(['PRESENT', 'ABSENT']),
  permissions: z.array(DatabaseUserPermission),
});

export function registerDatabaseClusterTools(server: McpServer, client: GraphQLClient): void {
  server.registerTool(
    'nexaa_db_cluster_list',
    { description: 'List all cloud database clusters' },
    async () => {
      const data = await client.request<GetCloudDatabaseClustersQuery>(
        GetCloudDatabaseClustersDocument,
      );
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.cloudDatabaseClusters, null, 2),
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_db_cluster_get',
    {
      description: 'Get a cloud database cluster by name and namespace',
      inputSchema: {
        name: z.string(),
        namespace: z.string(),
      },
    },
    async ({ name, namespace }) => {
      const data = await client.request<GetCloudDatabaseClusterQuery>(
        GetCloudDatabaseClusterDocument,
        { cloudDatabaseClusterInput: { name, namespace } },
      );
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.cloudDatabaseCluster, null, 2),
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_db_cluster_create',
    {
      description:
        'Create a cloud database cluster. Use nexaa_db_cluster_list_plans and nexaa_db_cluster_list_versions to discover valid plan IDs and spec values.',
      inputSchema: {
        name: z.string(),
        namespace: z.string(),
        plan: z.string().describe('Plan ID from nexaa_db_cluster_list_plans'),
        spec: z.object({
          type: z.string().describe('Database engine e.g. postgresql'),
          version: z.string().describe('Engine version e.g. 15'),
        }),
        databases: z.array(DatabaseInput).default([]),
        users: z.array(DatabaseUserInput).default([]),
        advisor: z
          .object({ enabled: z.boolean() })
          .optional()
          .describe('Enable database advisor (missing index suggestions)'),
        externalConnection: ExternalConnection.optional(),
      },
    },
    async (input) => {
      const data = await client.request<CloudDatabaseClusterCreateMutation>(
        CloudDatabaseClusterCreateDocument,
        { cloudDatabaseClusterInput: input },
      );
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.cloudDatabaseClusterCreate, null, 2),
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_db_cluster_modify',
    {
      description:
        'Modify a cloud database cluster. name and namespace are required; all other fields are optional.',
      inputSchema: {
        name: z.string(),
        namespace: z.string(),
        plan: z.string().optional(),
        databases: z.array(DatabaseInput).optional(),
        users: z.array(DatabaseUserInput).optional(),
        advisor: z.object({ enabled: z.boolean() }).optional(),
        externalConnection: ExternalConnection.optional(),
      },
    },
    async (input) => {
      const data = await client.request<CloudDatabaseClusterModifyMutation>(
        CloudDatabaseClusterModifyDocument,
        { cloudDatabaseClusterModifyInput: input },
      );
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.cloudDatabaseClusterModify, null, 2),
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_db_cluster_delete',
    {
      description:
        'Delete a cloud database cluster. WARNING: This permanently deletes the cluster and all its data. Requires confirm: true.',
      inputSchema: {
        name: z.string(),
        namespace: z.string(),
        confirm: z.boolean().describe('Must be true to confirm deletion'),
      },
    },
    async ({ name, namespace, confirm }) => {
      if (!confirm) {
        return {
          content: [
            {
              type: 'text',
              text: 'Deletion cancelled: confirm must be true to delete a database cluster.',
            },
          ],
        };
      }
      await client.request(CloudDatabaseClusterDeleteDocument, {
        cloudDatabaseClusterResourceInput: { name, namespace },
      });
      return {
        content: [{ type: 'text', text: `Database cluster "${name}" deleted.` }],
      };
    },
  );

  server.registerTool(
    'nexaa_db_cluster_list_plans',
    {
      description:
        'List available cloud database cluster plans with CPU, memory, storage, and pricing information',
    },
    async () => {
      const data = await client.request<ClusterPlansQuery>(ClusterPlansDocument);
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.cloudDatabaseClusterPlans, null, 2),
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_db_cluster_list_versions',
    {
      description:
        'List supported database engine types and versions (e.g. postgresql 15, mysql 8)',
    },
    async () => {
      const data = await client.request<ClusterVersionsQuery>(ClusterVersionsDocument);
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.cloudDatabaseClusterVersions, null, 2),
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_db_user_get_credentials',
    {
      description:
        'Retrieve the DSN (connection string) for a database user. The DSN contains credentials — treat it as sensitive.',
      inputSchema: {
        clusterName: z.string(),
        clusterNamespace: z.string(),
        username: z.string(),
      },
    },
    async ({ clusterName, clusterNamespace, username }) => {
      const data = await client.request<GetCloudDatabaseClusterUserCredentialsQuery>(
        GetCloudDatabaseClusterUserCredentialsDocument,
        {
          cloudDatabase: { name: clusterName, namespace: clusterNamespace },
          userName: username,
        },
      );
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.cloudDatabaseClusterUserCredentials, null, 2),
          },
        ],
      };
    },
  );
}

File: tools/database_cluster.go
package tools

import (
	"context"
	"fmt"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"github.com/nexaa-cloud/nexaa-cli/api"
)

var dbUserPermItemSchema = map[string]any{
	"type": "object",
	"properties": map[string]any{
		"databaseName": map[string]any{"type": "string"},
		"permission":   map[string]any{"type": "string", "enum": []string{"READ_ONLY", "READ_WRITE"}},
		"state":        map[string]any{"type": "string", "enum": []string{"PRESENT", "ABSENT"}},
	},
	"required": []string{"databaseName", "permission", "state"},
}

var dbInputItemSchema = map[string]any{
	"type": "object",
	"properties": map[string]any{
		"name":        map[string]any{"type": "string"},
		"state":       map[string]any{"type": "string", "enum": []string{"PRESENT", "ABSENT"}},
		"description": map[string]any{"type": "string"},
	},
	"required": []string{"name", "state"},
}

var dbUserInputItemSchema = map[string]any{
	"type": "object",
	"properties": map[string]any{
		"name":        map[string]any{"type": "string"},
		"password":    map[string]any{"type": "string"},
		"state":       map[string]any{"type": "string", "enum": []string{"PRESENT", "ABSENT"}},
		"permissions": map[string]any{"type": "array", "items": dbUserPermItemSchema},
	},
	"required": []string{"name", "state", "permissions"},
}

type dbInputArg struct {
	Name        string  `json:"name"`
	State       string  `json:"state"`
	Description *string `json:"description"`
}

type dbUserPermArg struct {
	DatabaseName string `json:"databaseName"`
	Permission   string `json:"permission"`
	State        string `json:"state"`
}

type dbUserInputArg struct {
	Name        string          `json:"name"`
	Password    *string         `json:"password"`
	State       string          `json:"state"`
	Permissions []dbUserPermArg `json:"permissions"`
}

func toDatabaseInputs(args []dbInputArg) []api.DatabaseInput {
	out := make([]api.DatabaseInput, len(args))
	for i, a := range args {
		out[i] = api.DatabaseInput{
			Name:        a.Name,
			State:       api.State(a.State),
			Description: a.Description,
		}
	}
	return out
}

// toDatabaseUserPermissions converts the MCP-facing permission args into
// nexaa-cli's input type. Shared by nexaa_db_user_create/modify and
// toDatabaseUserInputs so the conversion only needs to be updated in one
// place if api.DatabaseUserPermissionInput's shape ever changes.
func toDatabaseUserPermissions(args []dbUserPermArg) []api.DatabaseUserPermissionInput {
	out := make([]api.DatabaseUserPermissionInput, len(args))
	for i, a := range args {
		out[i] = api.DatabaseUserPermissionInput{
			DatabaseName: a.DatabaseName,
			Permission:   api.DatabasePermission(a.Permission),
			State:        api.State(a.State),
		}
	}
	return out
}

func toDatabaseUserInputs(args []dbUserInputArg) []api.DatabaseUserInput {
	out := make([]api.DatabaseUserInput, len(args))
	for i, a := range args {
		out[i] = api.DatabaseUserInput{
			Name:        a.Name,
			Password:    a.Password,
			State:       api.State(a.State),
			Permissions: toDatabaseUserPermissions(a.Permissions),
		}
	}
	return out
}

func RegisterDatabaseCluster(s *server.MCPServer) {
	s.AddTool(
		mcp.NewTool("nexaa_db_cluster_list",
			mcp.WithDescription("List all cloud database clusters"),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			result, err := withAPI(func(c *api.Client) ([]api.CloudDatabaseClusterResult, error) {
				return c.CloudDatabaseClusterList()
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_db_cluster_get",
			mcp.WithDescription("Get a cloud database cluster by name and namespace"),
			mcp.WithString("name", mcp.Required()),
			mcp.WithString("namespace", mcp.Required()),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			result, err := withAPI(func(c *api.Client) (api.CloudDatabaseClusterResult, error) {
				return c.CloudDatabaseClusterGet(api.CloudDatabaseClusterResourceInput{
					Name:      req.GetString("name", ""),
					Namespace: req.GetString("namespace", ""),
				})
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_db_cluster_create",
			mcp.WithDescription("Create a cloud database cluster. Use nexaa_db_cluster_list_plans and nexaa_db_cluster_list_versions to discover valid plan IDs and spec values."),
			mcp.WithString("name", mcp.Required()),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithString("plan", mcp.Required(), mcp.Description("Plan ID from nexaa_db_cluster_list_plans")),
			mcp.WithObject("spec", mcp.Required(), mcp.Description("Database engine spec"),
				mcp.Properties(map[string]any{
					"type":    map[string]any{"type": "string", "description": "e.g. postgresql"},
					"version": map[string]any{"type": "string", "description": "e.g. 15"},
				}),
			),
			mcp.WithArray("databases", mcp.Items(dbInputItemSchema)),
			mcp.WithArray("users", mcp.Items(dbUserInputItemSchema)),
			mcp.WithObject("externalConnection"),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			var args struct {
				Name               string           `json:"name"`
				Namespace          string           `json:"namespace"`
				Plan               string           `json:"plan"`
				Spec               struct {
					Type    string `json:"type"`
					Version string `json:"version"`
				} `json:"spec"`
				Databases          []dbInputArg     `json:"databases"`
				Users              []dbUserInputArg `json:"users"`
				ExternalConnection *extConnArg      `json:"externalConnection"`
			}
			if err := req.BindArguments(&args); err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			result, err := withAPI(func(c *api.Client) (api.CloudDatabaseClusterResult, error) {
				return c.CloudDatabaseClusterCreate(api.CloudDatabaseClusterCreateInput{
					Name:               args.Name,
					Namespace:          args.Namespace,
					Plan:               args.Plan,
					Spec:               api.CloudDatabaseClusterSpecInput{Type: args.Spec.Type, Version: args.Spec.Version},
					Databases:          toDatabaseInputs(args.Databases),
					Users:              toDatabaseUserInputs(args.Users),
					ExternalConnection: toExtConn(args.ExternalConnection),
				})
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_db_cluster_modify",
			mcp.WithDescription("Modify a cloud database cluster. name and namespace are required; all other fields are optional."),
			mcp.WithString("name", mcp.Required()),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithArray("databases", mcp.Items(dbInputItemSchema)),
			mcp.WithArray("users", mcp.Items(dbUserInputItemSchema)),
			mcp.WithObject("externalConnection"),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			var args struct {
				Name               string           `json:"name"`
				Namespace          string           `json:"namespace"`
				Databases          []dbInputArg     `json:"databases"`
				Users              []dbUserInputArg `json:"users"`
				ExternalConnection *extConnArg      `json:"externalConnection"`
			}
			if err := req.BindArguments(&args); err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			result, err := withAPI(func(c *api.Client) (api.CloudDatabaseClusterResult, error) {
				return c.CloudDatabaseClusterModify(api.CloudDatabaseClusterModifyInput{
					Name:               args.Name,
					Namespace:          args.Namespace,
					Databases:          toDatabaseInputs(args.Databases),
					Users:              toDatabaseUserInputs(args.Users),
					ExternalConnection: toExtConn(args.ExternalConnection),
				})
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_db_cluster_delete",
			mcp.WithDescription("Delete a cloud database cluster. WARNING: This permanently deletes the cluster and all its data. Requires confirm: true."),
			mcp.WithString("name", mcp.Required()),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithBoolean("confirm", mcp.Required(), mcp.Description("Must be true to confirm deletion")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			name := req.GetString("name", "")
			ns := req.GetString("namespace", "")
			if ok, cancelled := confirmDelete(req, "database cluster"); !ok {
				return cancelled, nil
			}
			_, err := withAPI(func(c *api.Client) (bool, error) {
				return c.CloudDatabaseClusterDelete(api.CloudDatabaseClusterResourceInput{Name: name, Namespace: ns})
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return mcp.NewToolResultText(fmt.Sprintf("Database cluster %q deleted.", name)), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_db_cluster_list_plans",
			mcp.WithDescription("List available cloud database cluster plans with CPU, memory, storage, and pricing information"),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			result, err := withAPI(func(c *api.Client) ([]api.CloudDatabaseClusterPlan, error) {
				return c.CloudDatabaseClusterListPlans()
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_db_cluster_list_versions",
			mcp.WithDescription("List supported database engine types and versions (e.g. postgresql 15, mysql 8)"),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			result, err := withAPI(func(c *api.Client) ([]api.CloudDatabaseClusterSpec, error) {
				return c.CloudDatabaseClusterListSpecs()
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_db_user_get_credentials",
			mcp.WithDescription("Retrieve the DSN (connection string) for a database user. The DSN contains credentials — treat it as sensitive."),
			mcp.WithString("clusterName", mcp.Required()),
			mcp.WithString("clusterNamespace", mcp.Required()),
			mcp.WithString("username", mcp.Required()),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			dsn, err := withAPI(func(c *api.Client) (string, error) {
				return c.CloudDatabaseClusterUserCredentials(
					api.CloudDatabaseClusterResourceInput{
						Name:      req.GetString("clusterName", ""),
						Namespace: req.GetString("clusterNamespace", ""),
					},
					req.GetString("username", ""),
				)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return mcp.NewToolResultText(dsn), nil
		},
	)
}
